tsl/src/remote/txn_resolve.c

```diff
@@ -107,10 +183,24 @@ remote_txn_heal_data_node(PG_FUNCTION_ARGS)
 	remote_result_close(res);
 
 	/*
-	 * Perform cleanup of all records if there are no unknown txns.
+	 * Perform cleanup of all records if there are no unknown txns and if the number of
+	 * resolved entities is same as the number of rows obtained from the datanode.
+	 *
+	 * In a heavily loaded system there's a possibility of ongoing transactions always being
+	 * present in which case we will never get a chance to clean up entries in "remote_txn"
+	 * table. So, we track healed gids in a list and delete those specific rows to keep the
+	 * "remote_txn" table from growing up indefinitely.
 	 */
-	if (list_length(unknown_txn_gid) == 0)
-		remote_txn_persistent_record_delete_for_data_node(foreign_server_oid);
+	if (list_length(unknown_txn_gids) == 0 && resolved == ntuples)
+		remote_txn_persistent_record_delete_for_data_node(foreign_server_oid, NULL);
+	else if (resolved)
+	{
+		ListCell *lc;
+		Assert(healed_txn_gids != NIL);
+
+		foreach (lc, healed_txn_gids)
+			remote_txn_persistent_record_delete_for_data_node(foreign_server_oid, lfirst(lc));
```

Comment:
Any way we can hit this in tests?

Reply:
I already added a test for this in `test_basic_persistent_record` function.